File: jvm/leetcode_scala/src/main/scala/leetcode/editor/cn/[538]把二叉搜索树转换为累加树.scala
package leetcode.editor.cn

object ConvertBstToGreaterTree {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) { var value: Int = _value var left: TreeNode = _left var right: TreeNode =
    * _right }
    */
  object Solution {
    import scala.util.control.TailCalls.*
    def convertBST(root: TreeNode): TreeNode =
      var sum = 0

      def traverse(node: TreeNode): TailRec[Unit] =
        node match
          case null => done(())
          case _ =>
            for
              _ <- traverse(node.right)
              _ = sum += node.value
              _ = node.value = sum
              _ <- traverse(node.left)
            yield ()

      traverse(root).result
      root

  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: scala/leetcode/src/main/scala/leetcode/editor/cn/[538]把二叉搜索树转换为累加树.kt
package leetcode.editor.cn

object ConvertBstToGreaterTree {

     class TreeNode(var `val`: Int) {
             var left: TreeNode? = null
             var right: TreeNode? = null
         }
//leetcode submit region begin(Prohibit modification and deletion)
    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     * class TreeNode(var `val`: Int) {
     *     var left: TreeNode? = null
     *     var right: TreeNode? = null
     * }
     */
    class Solution {
        var sum = 0
        fun convertBST(root: TreeNode?): TreeNode? {
            traverse(root)
            return root
        }

        fun traverse(node:TreeNode?) {
            if (node == null) return

            traverse(node.right)
            sum += node.`val`
            node.`val` = sum
            traverse(node.left)

        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}
